scripts/forger/seed.sh

```diff
@@ -0,0 +1,148 @@
+#!/bin/bash
+
+set -eEuo pipefail
+
+# This script is used to download and extract the blockchain data from a seed node.
+
+# Select datadir
+datadir="$HOME/.zen"
+
+# Detect network
+if grep -q "testnet=1" "$HOME/.zen/zen.conf" <(echo "${OPTS:-}"); then
+  datadir+="/testnet3"
+fi
+
+# Check if script needs to be run
+if [ "${FORCE_RESEED:-false}" = "true" ]; then
+  echo "FORCE_RESEED is true. Removing seed lock and complete files."
+  [ -f "${datadir}/.seed.lock" ] && rm -f "${datadir}/.seed.lock"
+  [ -f "${datadir}/.seed.complete" ] && rm -f "${datadir}/.seed.complete"
+  [ -d "${datadir}/blocks" ] && rm -rf "${datadir}/blocks"
+  [ -d "${datadir}/chainstate" ] && rm -rf "${datadir}/chainstate"
+  [ -f "${datadir}/*.tgz" ] && rm -f "${datadir}/*.tgz"
+elif [ -f "${datadir}/.seed.complete" ]; then
+  echo "'.seed.complete' file exists. Skipping seed step."
+  exit 0
+elif [ "${USE_SEED_FILE:-false}" = "false" ]; then
+  echo "USE_SEED_FILE is false. Skipping seed step."
+  touch "${datadir}/.seed.complete"
+  exit 0
+elif [ -d "${datadir}/blocks" ] && [ "$(ls -A "${datadir}"/blocks)" ]; then
+  echo "'blocks' directory exist and is not empty. Skipping seed step."
+  touch "${datadir}/.seed.complete"
+  exit 0
+elif [ -d "${datadir}/chainstate" ] && [ "$(ls -A "${datadir}"/chainstate)" ]; then
+  echo "'chainstate' directory exist and is not empty. Skipping seed step."
+  touch "${datadir}/.seed.complete"
+  exit 0
+fi
+
+# Functions
+download_and_extract_seed() {
+  local retries=3
+
+  # Create a lock file
+  touch "${datadir}/.seed.lock"
+
+  # Download loop
+  for ((i = 1; i <= retries; i++)); do
+    echo "Download attempt $i of $retries"
+
+    if ! aria2c -x 1 -s 1 -d "${datadir}" "${SEED_TAR_GZ_URL}" 2>/dev/null; then
```

Comment:
The advantage of aria2c is that it can greatly speed up DLs by using up to 16 parallel connections with `-x16 -s16`. Using `-x1 -s1` provides no advantage over using curl. Depending on the server we're downloading from this doesn't always work reliably. Cloudflare non-enterprise for example only allows up to 2 connections.

I wrote logic a few years ago in zen-fetch-params to handle these cases in addition to resume support of partially downloaded files. https://github.com/HorizenOfficial/zen/blob/f0c80a6853d2e77b39d019ac62fe74210e15a0aa/zcutil/fetch-params.sh#L31-L71